### PR TITLE
[sigh] Fix entitlements merging when changing team while resigning

### PR DESCRIPTION
**Pull request by kapfab**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Since fastlane 2.82.0, "advanced" entitlements can be addressed by `resign` with `--use_app_entitlements`option.
All is working fine when resigning without changing team, but there are issues with some of these entitlements when the team of the app to be resigned is different from the one of the new provisioning profile (which is a common use case for `resign`).
An example of this issue is mentioned by @heps and @aklowther on #11070.

These entitlements are the following:
- `com.apple.security.application-groups` (App Groups)
- `com.apple.developer.in-app-payments` (Apple Pay In-App Payments / Merchant IDs)
- `com.apple.developer.ubiquity-container-identifiers` (iCloud Containers)
- `com.apple.developer.icloud-container-identifiers` (iCloud Containers)
- `com.apple.developer.pass-type-identifiers` (Wallet / Pass Type IDs)

These 5 entitlements are linked to unique identifiers that are defined on the Developer Portal and that can't be shared between teams.
Hence, they must not be retrieved from the existing app if the team changes. In this case, they must be retrieved from the new provisioning profile instead.

There is only one potential issue with this : provisioning profiles only list a single generic value for pass types (even if the app entitlements allowed a restricted list of pass types), and we can't guess if a specific subset of pass types of the new team is to be used.
Allowing all Pass Types of one's team is common use and should be fine for most apps.
Anyway, there is no better way to do it for now without adding a bunch of parameters to `resign`.

### Description
The entitlements merging logic has been changed:
- The list of keys to be blacklisted from the provisioning profile has been updated and it now includes the specific keys above if there is no team change.
- The list of keys to be retrieved from the existing app has been updated and now it includes the specific keys above if there is no team change.
In case of team change, a warning is displayed to inform the user of these specific merging rules.

The logic for replacing old App/Team IDs by the new ones has been optimized to prevent 2*N calls to `sed` on the whole patched entitlements file (this was overkill and could have exceptionally led to unintentional replacements in previous entitlements' values).
The logic to populate the iCloud Environment entitlement value has also been integrated to this block for legibility.

Also, the regular expression used with `sed` for extracting the entitlements values from the app entitlements was not working as expected on the multiline XML result of `PlistBuddy -x`.
This was not a big issue as this complete XML result with headers was correctly read by `plutil -insert … -xml` but lines of XML were dumped in the log for each entitlement instead of a single XML element.
This is fixed and verbose logs are now more legible when using `--use_app_entitlements`.

### CODE REVIEW

**Review by mgrebenets — COMMENTED** [review]
Looks legit 👍

I want to give it a go in our CI job at work to see if it still resigns our app correctly.

Always good to give it a go like this for verification.

**Comment by kapfab** [comment]
I couldn’t agree more!

**Comment by joshdholtz** [comment]
This looks amazing but also way over my head 🙌 Do you have a good suggested way of testing this? I have actually never resigned anything before 🙃

**Comment by mgrebenets** [comment]
@joshdholtz 
I think we _may_ be able to have some tests.
That would involve a couple of dev accounts though.

E.g. we have an ipa signed with Dev A's AppStore dev account, and want to resign it to Enterprise version.
For resign operation we need to have a private key available as part of Mac OS X Keychain.
We can get p12 protected by a password or pull certs from repo using `match`.

The question is how do we secure that certificate (private key)?
Is there a way to store some password-protected assets as part of fastlane repo?

Alternatively, we make this private key available for everyone, which means anyone can pull it in and use it to sign some random iOS apps.

**Comment by aklowther** [comment]
I just tested this, it resolved the issues brought in with 2.82.0

**Comment by DanToml** [comment]
Assigning myself to take a look at this over the weekend, but if someone else gets to it first that is 👍 .

**Review by mgrebenets — APPROVED** [review]

File: sigh/lib/assets/resign.sh
Comment (getaaron) [inline sigh/lib/assets/resign.sh]: > WARNING: If these capabilities are enabled…

Can we check whether the capabilities are enabled and show a more tailored message? Broad, not-necessarily-applicable error messages sometimes confuse and scare devs :)

This could be as simple as printing the `.entitlements` file and telling them what to look for. I also opened Xcodeproj issue https://github.com/CocoaPods/Xcodeproj/issues/562 to make this easier to parse.

File: sigh/lib/assets/resign.sh
Comment (getaaron) [inline sigh/lib/assets/resign.sh]: Can we have one source array for both `ENTITLEMENTS_TRANSFER_RULES += ` and `BLACKLISTED_KEYS+=` above? I want to reduce the risk someone will accidentally modify one array in the future and forget to modify the other.

**Comment by getaaron** [comment]
@kapfab This looks great, thanks so much for the PR! I added two suggestions, let me know what you think?

**Comment by getaaron** [comment]
Also [the macOS CircleCI build](https://circleci.com/gh/fastlane/fastlane/17188) failed while they were having an outage, you may need to push another commit to get it to rerun.

File: sigh/lib/assets/resign.sh
Comment (kapfab) [reply sigh/lib/assets/resign.sh]: You're right, but that seemed a good compromise regarding the fact that this issue is preventing people from resigning production apps.

Regarding your last point, Xcodeproj files are not involved here as `--use_app_entitlements` retrieves the entitlements directly from the app binary.
Anyway, IMHO, if someone with no knowledge of the way entitlements work wants to use `resign` with `--use_app_entitlements` option while changing team, they need a tutorial instead of just a few lines in a log to tell them what to look for.

File: sigh/lib/assets/resign.sh
Comment (kapfab) [reply sigh/lib/assets/resign.sh]: I was willing to do that but:
- the `ENTITLEMENTS_TRANSFER_RULES` array can include patching rules (`TEAM_ID`, `APP_ID`, and now `ICLOUD_ENV`) that should then be ignored for blacklisting purposes
- we could have near-future cases where a key is not in the blacklist but still added to the transfer rules to be retrieved from the existing app if defined (used as a preferred value)

That mean we would have to maintain a single array with "behavior types" in addition to the potential patching rule(s) for each of the keys.

This should definitely be investigated (along with a whitelist approach), but this goes beyond this "quick fix".

**Comment by aklowther** [comment]
@kapfab ping

**Comment by janpio** [comment]
Restarted the tests on CircleCI, passes.

**Comment by janpio** [comment]
Nah @kapfab, more that I don't understand how it works and I keep trying to remove "needs-attention" although it uses the total age of the PR I think ;)

@getaaron Can you take a look at the comment replies from @kapfab above please? Thanks.

File: sigh/lib/assets/resign.sh
Comment (lyndsey-ferguson) [inline sigh/lib/assets/resign.sh]: I do not understand the entitlements transfer rules. Can it be improved here for people who are not as familiar with this?

**Review by lyndsey-ferguson — COMMENTED** [review]
This looks like a solid change. It would be great if the entitlements transfer rules block could have a comment that better explains how it works.

**Comment by aklowther** [comment]
@kapfab any chance this could get some attention soon?

**Review by getaaron — APPROVED** [review]
This LGTM but I don’t have a way to test it since I only have access to one signing account at the moment.

**Comment by janpio** [comment]
Ohoh, this now has merge conflicts - and the second one doesn't look trivial. Could you take a look @kapfab please?

**Comment by rodrigoelp** [comment]
Hi guys, I am joining the group of people affected by this.

In my particular case, we create a build that gets resigned twice: once for internal/enterprise release (for testing) and other one to the app store... these two belong to different teams and it means we need to transfer some of the entitlements over the new team id (and other identifiers).

I am trying to use the `--use-app-entitlements` in it but the output tends to be invalid because App Groups do not match.

Is there an estimated time to release? What's blocking this PR (other than the merge conflicts)?

**Comment by getaaron** [comment]
@rodrigoelp nothing blocking the PR besides merge conflicts and request for improved explanation here https://github.com/fastlane/fastlane/pull/12131#discussion_r237631944

**Comment by rodrigoelp** [comment]
That's great read @getaaron. Hopefully the PR will get through soon :)

**Comment by getaaron** [comment]
@rodrigoelp Feel free to take a stab at it yourself if you’re so inclined

**Comment by rodrigoelp** [comment]
> @rodrigoelp Feel free to take a stab at it yourself if you’re so inclined

I would love to but I don't know what to improve on the transfer rules and I haven't checked the merge clash. I can try the later tomorrow once I get to the office.

**Comment by kapfab** [comment]
> Is there an estimated time to release? What's blocking this PR (other than the merge conflicts)?
> 
> Ping @kapfab

Hi,

I spent hours if not days to identify and test all merge cases last year to deliver a working fix for everyone (a single entitlement was missing for our own needs).
It has been blocked for one year now.
Initially because of a CircleCI issue, then I was asked to document transfer rules that were not documented at all before my PR and now, as time goes, another PR did its way and brought a conflict.

Honestly, trying twice to fix things I did not break is not in my current top priorities.
Though, if someone has time to work on it, I’ll be happy to bring a quick help if needed.

**Comment by lyndsey-ferguson** [comment]
> then I was asked to document transfer rules that were not documented at all before my PR and now

As I don't understand how they work, for me it would be enough if you documented what you **do** know about the rules as you touched them.

**Comment by kapfab** [comment]
> As I don't understand how they work, for me it would be enough if you documented what you **do** know about the rules as you touched them.

As far as I can remember, I added a comment to explain each block I touched and I even added a comment for each entitlement, explaining why it has to be blacklisted from the PP.

I know how entitlement merging (and this shell script above all) is complicated to understand, maintain and test.
If you have any question, I'd be pleased to answer, but I just don't have time to re-document the whole process by commenting code.

**Comment by silvervest** [comment]
Any update on when this PR will be merged? We are being affected where resigning a package containing an extension and an app group, the app group is not being updated within the extension's entitlements. Have confirmed using this commit that it resolves this issue.

**Comment by janpio** [comment]
There are now 2 merge conflicts, one simpler and one more involved that would have to be fixed first :/

**Comment by silvervest** [comment]
Hmm I see. I forked, rebased to fastlane/fastlane:master and resolved the conflicts, per https://github.com/silvervest/fastlane/commit/8281d6caf273a513adef50cfc7ed2eb40e9eccb6

I'm not sure how to get this fed back into this PR though, if that's even possible?

**Comment by janpio** [comment]
In theorey you could create a PR to @kapfab's repo which they would have to merge. In practice it might be easier to create a new PR with the same changes (and link to it here) - on merge you will both get credit in the commit history automatically.

**Comment by silvervest** [comment]
Thanks for the advice @janpio - tried going the in-theory PR into @kapfab's repo, but the PR would include all changes between master then and now, so I went the in-practice option and submitted https://github.com/fastlane/fastlane/pull/15380

**Comment by janpio** [comment]
Closing this as #15388 will replace it - the original author @kapfab will also get credit if that PR is merged.

**Comment by janpio** [comment]
Hey everyone here, would be great if you could have a look at https://github.com/fastlane/fastlane/pull/15388 which replaced this PR and leave an "approve" review if you think this can be merged 👍

**Pull request closed without merging**
